### PR TITLE
Add an example for a benchmark with an `Activity` that is not exported.

### PR DESCRIPTION
**Pull request by tikurahul**
Test:

```
$ adb shell am instrument -w -m    -e debug false -e class 'com.example.macrobenchmark.NonExportedActivityBenchmark#start' com.example.macrobenchmark.test/androidx.test.runner.AndroidJUnitRunner
Connected to process 26738 on device 'google-pixel_5-06131FDD4000FF'.
benchmark: NonExportedActivityBenchmark_start
benchmark:   frameTime50thPercentileMs   min   5,   median   5,   max   5
benchmark:   frameTime90thPercentileMs   min   5,   median   5,   max   5
benchmark:   frameTime95thPercentileMs   min   5,   median   6,   max   7
benchmark:   frameTime99thPercentileMs   min   7,   median   9,   max  10
benchmark:             totalFrameCount   min 584,   median 588,   max 591
benchmark:
```

### CODE REVIEW

File: MacrobenchmarkSample/macrobenchmark/src/androidTest/java/com/example/macrobenchmark/NonExportedActivityBenchmark.kt [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+        private const val PACKAGE_NAME = "com.example.macrobenchmark.target"
+        private const val RECYCLER_ACTIVITY_CLASS = "$PACKAGE_NAME.NonExportedRecyclerActivity"
+        private const val ACTION = "$PACKAGE_NAME.ACTIVITY_LAUNCHER_ACTIVITY"
+        private const val LAUNCH_RESOURCE_ID = "launchRecyclerActivity"
+        private const val RECYCLER_RESOURCE_ID = "recycler"
+
+        /**
+         * Waits until an [android.app.Activity] with the given `className` is visible.
+         */
+        fun UiDevice.waitUntilActivity(className: String) {
```
Comment (ChrisCraik) [inline MacrobenchmarkSample/macrobenchmark/src/androidTest/java/com/example/macrobenchmark/NonExportedActivityBenchmark.kt:94]: In my testing, `By.clazz()` only looks for view classes, not activity classes. Can you confirm it's not just hitting timeout?

What you can do is just what we do here: https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:benchmark/macro/src/main/java/androidx/benchmark/macro/MacrobenchmarkScope.kt;l=102;drc=3c2bd027309bfdbd148967952355db805c0d7db7

My understanding is simply by using `Until.hasObject(By.pkg(packageName).depth(0))`, it will wait for an activity to be ready of the package name to be ready to display, so will wait for its inflation to complete.

File: MacrobenchmarkSample/macrobenchmark/src/androidTest/java/com/example/macrobenchmark/NonExportedActivityBenchmark.kt [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+import java.util.concurrent.TimeUnit
+
+@LargeTest
+@SdkSuppress(minSdkVersion = 29)
+@RunWith(AndroidJUnit4::class)
+class NonExportedActivityBenchmark {
+    @get:Rule
+    val benchmarkRule = MacrobenchmarkRule()
+
+    @Test
```
Comment (ChrisCraik) [inline MacrobenchmarkSample/macrobenchmark/src/androidTest/java/com/example/macrobenchmark/NonExportedActivityBenchmark.kt:43]: Add a comment:

This benchmark measures the performance of an activity that isn't exported in the manifest, and thus can't be directly launched. To measure it, we use UiAutomator to first navigate to the destination in the setupBlock.

File: MacrobenchmarkSample/macrobenchmark/src/androidTest/java/com/example/macrobenchmark/NonExportedActivityBenchmark.kt [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+            // Try switching to different compilation modes to see the effect
+            // it has on frame timing metrics.
+            compilationMode = CompilationMode.None,
+            iterations = 3,
+            setupBlock = {
+                // Before starting to measure, navigate to the UI to be measured
+                val intent = Intent()
+                intent.action = ACTION
+                startActivityAndWait(intent)
+                val launchRecyclerActivity = device.findObject(
```
Comment (ChrisCraik) [inline MacrobenchmarkSample/macrobenchmark/src/androidTest/java/com/example/macrobenchmark/NonExportedActivityBenchmark.kt:59]: Add a comment:

// click a button to launch the target activity. While we use resourceId here to find the button, you could also use accessibility info or button text content.

File: MacrobenchmarkSample/macrobenchmark/src/androidTest/java/com/example/macrobenchmark/NonExportedActivityBenchmark.kt [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+            )
+            // Set gesture margin to avoid triggering gesture navigation
+            // with input events from automation.
+            recycler.setGestureMargin(device.displayWidth / 5)
+            for (i in 1..10) {
+                recycler.scroll(Direction.DOWN, 2f)
+                device.waitForIdle()
+            }
+            // Mark activity as finished
+            device.pressBack()
```
Comment (ChrisCraik) [inline MacrobenchmarkSample/macrobenchmark/src/androidTest/java/com/example/macrobenchmark/NonExportedActivityBenchmark.kt:80]: why is this necessary?

File: MacrobenchmarkSample/macrobenchmark/src/androidTest/java/com/example/macrobenchmark/NonExportedActivityBenchmark.kt [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+            packageName = PACKAGE_NAME,
+            metrics = listOf(FrameTimingMetric()),
+            // Try switching to different compilation modes to see the effect
+            // it has on frame timing metrics.
+            compilationMode = CompilationMode.None,
+            iterations = 3,
+            setupBlock = {
+                // Before starting to measure, navigate to the UI to be measured
+                val intent = Intent()
+                intent.action = ACTION
```
Comment (ChrisCraik) [inline MacrobenchmarkSample/macrobenchmark/src/androidTest/java/com/example/macrobenchmark/NonExportedActivityBenchmark.kt:57]: add Intent.FLAG_ACTIVITY_CLEAR_TASK and Intent.FLAG_ACTIVITY_NEW_TASK, with a comment saying this ensures each measurement is with a unique activity

I think this also means you don't need to nav back anymore

File: MacrobenchmarkSample/macrobenchmark/src/androidTest/java/com/example/macrobenchmark/NonExportedActivityBenchmark.kt [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+@LargeTest
+@SdkSuppress(minSdkVersion = 29)
+@RunWith(AndroidJUnit4::class)
+class NonExportedActivityBenchmark {
+    @get:Rule
+    val benchmarkRule = MacrobenchmarkRule()
+
+    @Test
+    fun start() {
```
Comment (ChrisCraik) [inline MacrobenchmarkSample/macrobenchmark/src/androidTest/java/com/example/macrobenchmark/NonExportedActivityBenchmark.kt:44]: scroll?

File: MacrobenchmarkSample/app/src/main/AndroidManifest.xml [hunk 7]
```diff
@@ -50,6 +50,18 @@
...
+        <activity
+            android:name=".ActivityLauncherActivity"
+            android:exported="true">
+            <intent-filter>
+                <action
+                    android:name="com.example.macrobenchmark.target.ACTIVITY_LAUNCHER_ACTIVITY" />
+                <category android:name="android.intent.category.DEFAULT" />
+            </intent-filter>
+        </activity>
+        <!-- non-exported activities -->
```
Comment (ChrisCraik) [inline MacrobenchmarkSample/app/src/main/AndroidManifest.xml:62]: change comment:

Non exported activity, with no action can be launched internally, but not directly by a macrobenchmark. See NonExportedActivityBenchmark for how to navigate there indirectly.

File: MacrobenchmarkSample/macrobenchmark/src/androidTest/java/com/example/macrobenchmark/NonExportedActivityBenchmark.kt (hunk 4 above)
Comment (tikurahul) [reply MacrobenchmarkSample/macrobenchmark/src/androidTest/java/com/example/macrobenchmark/NonExportedActivityBenchmark.kt]: Its necessary if we don't use `NEW_TASK | CLEAR_TASK`. I am using it now.

File: MacrobenchmarkSample/macrobenchmark/src/androidTest/java/com/example/macrobenchmark/NonExportedActivityBenchmark.kt (hunk 1 above)
Comment (tikurahul) [reply MacrobenchmarkSample/macrobenchmark/src/androidTest/java/com/example/macrobenchmark/NonExportedActivityBenchmark.kt:97]: I have a 10 second timeout, and it definitely does not for wait for that long. 😕 

`Until.hasObject(By.pkg(packageName).depth(0))` may not be the right thing given we already have a visible activity with the same package name right ?

Comment (ChrisCraik) [reply MacrobenchmarkSample/macrobenchmark/src/androidTest/java/com/example/macrobenchmark/NonExportedActivityBenchmark.kt:97]: > given we already have a visible activity

That's what my initial thinking, but in my testing hasObject() would always pause until the in-progress launch completed.


Can you confirm that the activity arg you pass is used? That is, you will wait for the 10 second timeout if you pass a different activity string? I'm pretty sure I tested that behavior, and it didn't work with fully qualified activity names. I tried manually passing a string, as well as querying intent.getComponent().name (or whatever it is).

On that note, be sure to throw if the wait(hasObject(...)) returns false, so there's no ambiguity whether it found the class.

Comment (tikurahul) [reply MacrobenchmarkSample/macrobenchmark/src/androidTest/java/com/example/macrobenchmark/NonExportedActivityBenchmark.kt:97]: Yes, there is a 10 second delay, if i refer to a class that does not exist. So it looks like the name is definitely being used. 

> On that note, be sure to throw if the wait(hasObject(...)) returns false, so there's no ambiguity whether it found the class.
That should happen already, because `findObject(...)` returns `null` and will result in an NPE.
